rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -86,22 +86,30 @@ Player::Player(const std::string & node_name, const rclcpp::NodeOptions & node_o
 }
 
 Player::Player(
-  std::shared_ptr<rosbag2_cpp::Reader> reader,
+  std::unique_ptr<rosbag2_cpp::Reader> reader,
   const rosbag2_storage::StorageOptions & storage_options,
   const rosbag2_transport::PlayOptions & play_options,
   const std::string & node_name,
   const rclcpp::NodeOptions & node_options)
 : rclcpp::Node(
     node_name,
     rclcpp::NodeOptions(node_options).arguments(play_options.topic_remapping_options)),
-  reader_(reader),
+  reader_(std::move(reader)),
   storage_options_(storage_options),
   play_options_(play_options)
 {}
 
 Player::~Player()
 {
-  // reader_->reset();
+  if (reader_) {
+    reader_->reset();
+  }
+}
+
+rosbag2_cpp::Reader * Player::release_reader()
+{
+  reader_->reset();
```

Comment:
This isn't `.reset`, it's `->reset`, which is calling the reset method of the object itself, not the reset method for the smart pointer.

Reply:
Sorry I missed that "This isn't .reset, it's ->reset". \
But it is still unclear for me why do we need to reset underlying storage in reader before releasing pointer to the reader?

Reply:
This makes it so that the reader is not in an intermediate state, like halfway through the bag. The new owner can call `open` successfully after this.

Reply:
Ok good. now it is clear for me. \
While it's arguable whether it should be reset  or not, at least I wouldn't expect it to be reset on `release_reader` call. \
I think at least we will need to add comment about it somewhere, perhaps in doxygen comment for the `Player::release_reader()` method.

Reply:
@emersonknapp @Karsten1987 BTW. This method `Player::release_reader()` looks very unsafe. What if someone will call it somewhere in the middle of the playback?

It would be more safely to force "`stop`" for player before releasing reader from it.

Reply:
At this point I think I will default to "PRs welcome" instead of continuing to discuss minor points on merged PRs. If you can provide a test proving the unsafeness of the current behavior, all the better.

Reply:
@emersonknapp I didn't expect that someone will take my comments as an action items for fixing already merged PR's. \
The purpose of my comments is a discussion to eliminate potential issues and come to the common opinion about findings and perhaps introduce some improvements in future if we would agree up on it.

Reply:
I think that this intention is good. In practice, a merged PR is not a very good place to track future work, though. A PR or an enhancement ticket are more likely to move forward - this conversation is likely to get forgotten.

On open source projects, the most important motion comes in the form of contributions. Discussion is fine but if nobody commits to action, it doesn't usually amount to anything. So that's my opinion on this particular conversation - that it would be more productively continued in the form of a code review or backlog ticket.

Reply:
@emersonknapp I agree that separate issue or pull request would be more appropriate place for discussion. I've just have some many things to do that I don't have time for it. And if I will not write my thoughts somewhere It will be forgotten completely. At least here is a PR where relevant  changes was introduced. I will try to raise a new issue for discussion next time.